### ISSUE TITLE
SPIKE: Flutter Web Applications

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-flutter#9]
One of the _major_ selling points of Flutter is that it's built with Dart. 
That means it can be compiled to _any_ target platform (_including the web!_)
Starting earlier in 2019 there has been ("_beta_") support for Web in Flutter: https://flutter.dev/web
![flutter-web](https://user-images.githubusercontent.com/194400/71715277-d23c3a80-2e08-11ea-8688-47f14cb0494f.png)

Official getting started guide: https://flutter.dev/docs/get-started/web

> @Nash0x7e2 wrote this overview of "Ins and Outs of Flutter Web":
https://medium.com/flutter-community/ins-and-outs-of-flutter-web-7a82721dc19a
> @zubairehman has written this great post on getting started building a Web App with Flutter:
https://itnext.io/flutter-for-web-c75011a41956 
> @mikegazzaruso wrote Flutter For Web: a concrete example on flutter-community:
https://medium.com/flutter-community/flutter-for-web-a-concrete-example-4ed293545008
> Sample web app: https://github.com/kenreilly/flutter-web-example

# Todo 

+ [ ] install Flutter on `localhost` 
+ [ ] Follow the official instructions to build a "hello world" web application
+ [ ] Deploy it to a conventional web hosting service e.g. `Fly.io` to test the process.

[![image](https://user-images.githubusercontent.com/194400/72090247-92dc8500-3305-11ea-9b3b-8fed09099f40.png)](https://youtu.be/tot-R_q5-0o?t=6)
https://youtu.be/tot-R_q5-0o?t=6

**Comment by nelsonic** [comment]
@iteles I get the feeling that this one will be _particularly_ effective for you to learn Flutter Web: 😉 
[![image](https://user-images.githubusercontent.com/194400/72090634-5c533a00-3306-11ea-93e4-80d772bf42aa.png)](https://youtu.be/33kyEzDMTZU)

**Comment by nelsonic** [comment]
Java ... 🤦‍♂ 
![image](https://user-images.githubusercontent.com/194400/72091817-f4522300-3308-11ea-815e-70ecddbc80b8.png)

**Comment by nelsonic** [comment]
```
Failed to download the Android dependencies
Downloading Android Maven dependencies...                           0.4s
```

**Comment by nelsonic** [comment]
```
flutter doctor
```
![image](https://user-images.githubusercontent.com/194400/72092232-cfaa7b00-3309-11ea-91f3-227f02d35517.png)

We see the following warning:
"idevice_id" cannot be opened because the developer cannot be verified
![image](https://user-images.githubusercontent.com/194400/72092199-befa0500-3309-11ea-8524-43e59e37e036.png)

Known issue: https://github.com/flutter/flutter/issues/42969 reading ... 👀

**Comment by SimonLab** [comment]
Almost there on my side (Ubuntu 19.10)
![image](https://user-images.githubusercontent.com/6057298/72103229-ae08be00-3320-11ea-8d74-b5e3b60b2d0c.png)

**Comment by SimonLab** [comment]
I couldn't connect my phone to my computer (I think it's an issue with my phone or ubuntu, not flutter directly), however I manage to create a virtual device:
![image](https://user-images.githubusercontent.com/6057298/72107355-0a6fdb80-3329-11ea-9647-134b43f22e2f.png)

**Comment by miguelmartins17** [comment]
@nelsonic I was Following this tutorial to install Flutter on Mac:https://flutter.dev/docs/get-started/install/macos 
But I made the exactly things they say here and "flutter --version" says flutter its not recognised even when installed.

**Comment by miguelmartins17** [comment]
@SimonLab Were you able to connect the phone to Android Studio or not yet?
I had the same problem when I started using Android Studio.
Here is a link that might help:https://developer.android.com/training/basics/firstapp/running-app#RealDevice

**Comment by nelsonic** [comment]
Lots more things to install apparently, so I will need to pay a visit to the local cafe with free Wifi ... ⏳ 
```
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-GB)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_HOME to that location.
      You may also want to add it to your PATH environment variable.

[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install:
        sudo gem install cocoapods
[!] Android Studio (not installed)
[!] VS Code (version 1.41.1)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[!] Connected device
    ! No devices available

! Doctor found issues in 5 categories.
```

**Comment by SimonLab** [comment]
@miguelmartins17 thanks for the link. I think the issue is linked to my Ubuntu not mounting/detecting my phone automatically. still trying to find a solution for this.

**Comment by SimonLab** [comment]
ok I think I'm good to go now:
![image](https://user-images.githubusercontent.com/6057298/72148815-a2a5a900-3399-11ea-9afb-ba89466ab2b8.png)

The issue wasn't with Ubuntu or my phone at the end but was due to the wrong type of usb cable I was using. Some usb cables are only used for charging devices, I had to find one that I can use to transfer data 
see https://electronics.stackexchange.com/questions/140225/how-can-i-tell-charge-only-usb-cables-from-usb-data-cables

**Comment by SimonLab** [comment]
Testing simple android app on my phone. It's true that I was impress by the hot reload, the project took a bit of time to compile (2 or 3 minutes) on my computer but once it was done I manage to quickly change the colour and the text easily.

<img src="https://user-images.githubusercontent.com/6057298/72156826-5e70d380-33ae-11ea-8b6f-4dddf3738e04.png" height="500">

**Comment by SimonLab** [comment]
For creating a web application we need to use the `beta` channel of flutter, https://flutter.dev/docs/get-started/web
![image](https://user-images.githubusercontent.com/6057298/72161241-a811ec00-33b7-11ea-876d-befd4e24b91a.png)

```sh
flutter channel beta
flutter upgrade
flutter config --enable-web
```

![image](https://user-images.githubusercontent.com/6057298/72161457-1060cd80-33b8-11ea-8981-2799f49546d1.png)

Now that the the web option for flutter is enable we can check the list of devices with:
```sh
flutter devices
```
I still have my phone connected but we can now also see 2 more devices: chrome and a web server
![image](https://user-images.githubusercontent.com/6057298/72161604-5584ff80-33b8-11ea-9887-6a5ed503016f.png)

We can also run `flutter doctor` on VScode to see which devices are connected:
![image](https://user-images.githubusercontent.com/6057298/72161822-b9a7c380-33b8-11ea-8d3f-e988f870a1cb.png)

I had to compile my current flutter project for the web with the command line`flutter create .` (I this point I didn't try to run the project directly with VScode but I suspect that the IDE would have done this for me). When ready the application is lauched on a chrome instance: 
![image](https://user-images.githubusercontent.com/6057298/72162288-9d585680-33b9-11ea-8a3b-d971788cd1f3.png)

I've noticed however that the state is not kept for the web app when modifications are applied to the code, I'm not sure if this due to the beta or if it's specific to the web app.

To recap once all the tools are installed it's pretty nice to be able to run the application on a real device, emulator and on chrome with only one code base. The hot reload for real device and chrome (expect the state which is not save on chrome) is working really well, for the emulator it was a bit mroe slow on my computer. Now we need to see if we can build more complex application with Flutter

**Comment by nelsonic** [comment]
Android studio is a _beast_! 
![image](https://user-images.githubusercontent.com/194400/72170817-773ab280-33c9-11ea-8a98-faab6cac63d2.png)

![image](https://user-images.githubusercontent.com/194400/72170801-6f7b0e00-33c9-11ea-943e-e66a46ab3ce7.png)

Going to watch this tutorial over the weekend:
[![image](https://user-images.githubusercontent.com/194400/72170969-cb459700-33c9-11ea-929b-de0a66c2395c.png)](https://youtu.be/dc3B_mMrZ-Q)
https://youtu.be/dc3B_mMrZ-Q

**Comment by miguelmartins17** [comment]
In my opinion for simple and interactive Android applications Android Studio is one of the best.

**Comment by miguelmartins17** [comment]
Of course, we can also make more complex applications with Android but right now I'm very interested in Flutter too.

**Comment by nelsonic** [comment]
With the release of **`Flutter 2.0`** Web Support is now "stable" (fully supported!)
[![image](https://user-images.githubusercontent.com/194400/111999657-17837f80-8b15-11eb-9ff3-0a82180f8f75.png)](https://youtu.be/HAstl_NkXl0)

**Comment by nelsonic** [comment]
https://betterprogramming.pub/flutter-is-about-to-win-over-the-web-be0a205af03d

**Comment by nelsonic** [comment]
We need to figure out how to deploy/host a Web App built with `Flutter` and see how we can optimise it.
Is the "bundle" going to be too big to be performant on a more basic device like an older `iPhone` or `Android` phone?

@LuchoTurtle can you please look into this? 
100% of the people ***trying*** our `App` for the first time will not download and install it.
We need the ***`PWA`*** experience to be as good as the `Native` experience (within reason).

> **Note**: not saying that _longer_-term people won't download and install the `Native` mobile app.
Just that in the beginning, getting people to download an `App` _before_ they can `try` it is a non-starter. 

Please confirm you understand what is required here. 🙏  (leave a comment on this issue).

**Comment by LuchoTurtle** [comment]
@nelsonic do you want me to create a simple todo web app in Flutter and deploy it? 
I can create a PR and a folder in this repo dedicated just to the web app. 
After implementing, we can take a look at Lighthouse insights and bundle sizes. 

If I'm correct, just give me the heads up 👍

**Comment by nelsonic** [comment]
Yes, please. 
A simple Todo App is a good starting point for testing a `Flutter PWA`. 👍 
But also, please take a look at: https://github.com/dwyl/flutter-phoenix-channels-demo/issues/1
If you can deploy that micro-app to `Fly.io` it would showcase `Flutter` connecting to `Phoenix`. 👌

**Comment by LuchoTurtle** [comment]
In that case, I'll just use the mobile todo app from https://github.com/dwyl/flutter-todo-list-tutorial/pull/28 (it's not yet merged but it's workable and has been the foundation for https://github.com/dwyl/supabase-flutter-demo, for example)
and work on https://github.com/dwyl/flutter-phoenix-channels-demo/issues/1 first.

https://github.com/dwyl/flutter-todo-list-tutorial/pull/28 already has Phoenix connecting to a Phoenix API. But I wager you meant having a websocket connection instead of RESTful requests.

**Comment by nelsonic** [comment]
https://github.com/dwyl/flutter-todo-list-tutorial/pull/28 merged. please proceed. 🙏 
WebSockets are the superpower of `Elixir` & `Phoenix`. 
Having a `REST` API is incidental (but obvs required for people to run their own analysis etc.)
Thanks.

**Comment by nelsonic** [comment]
`Flutter Web` Apps are good enough that we can now use `Flutter` to build our Web App with PWA support. Closing. ✅
